### PR TITLE
feat: Add support for asynchronous invocation of `FilesAdapter.getFileLocation`

### LINKED ISSUE #9268: Add support for async `FilesAdapter.getFileLocation`
### New Feature / Enhancement Checklist

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Before posting search [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

The `FilesAdapter` protocol expects `getFileLocation` to be a synchronous method.

### Feature / Enhancement Description

Add support for `getFileLocation` to be either an asynchronous or synchronous method:
- Support for asynchronous is required for file adapters that generate the file URI asynchronously, for example the S3 file adapter using a pre-signed URL, see https://github.com/parse-community/parse-server-s3-adapter/issues/197.
- Support for synchronous is required for backward compatibility. 

### Example Use Case

See https://github.com/parse-community/parse-server-s3-adapter/issues/197.

### PR DESCRIPTION
**Pull request by vahidalizad**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue
Closes: #9268

## Approach
Added async support to `FilesAdapter.getFileLocation`.

## Tasks
- [x] Added async support to `FilesAdapter.getFileLocation`.

### CODE REVIEW

File: spec/FilesController.spec.js [hunk 1]
```diff
@@ -21,11 +21,11 @@ const mockAdapter = {
 
 // Small additional tests to improve overall coverage
 describe('FilesController', () => {
-  it('should properly expand objects', done => {
+  it('should properly expand objects', async done => {
```
Comment (mtrezza) [inline spec/FilesController.spec.js]: Could you keep the sync test and add the async test as an addition test? We need to test whether both work. Also it seems there are other tests failing, see CI.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: spec/FilesController.spec.js (hunk 1 above)
Comment (vahidalizad) [reply spec/FilesController.spec.js]: Yeah, sure! I've added a new test to check getFileLocation for both the async and sync functions. I believe I've also fixed the errors that were causing the CI failures. Could you please review it and let me know if everything looks good now?

**Comment by vahidalizad** [comment]
I haven't made any changes to the logic files just added tests for more coverage and the previous run was successful for all tests. Now, we're seeing failing CIs outside of the ParseObject tests. Can anyone help me understand why this might be happening?

**Comment by mtrezza** [comment]
The tests are now randomized, that may be a reason. If they are unrelated to this PR no worries, they may just be flaky. I'll rerun the CI, and they should pass at some point.

File: spec/FilesController.spec.js [hunk 2]
```diff
@@ -21,11 +21,14 @@ const mockAdapter = {
...
-  it('should properly expand objects', done => {
+  it('should properly expand objects with sync getFileLocation', async () => {
     const config = Config.get(Parse.applicationId);
     const gridFSAdapter = new GridFSBucketAdapter('mongodb://localhost:27017/parse');
+    gridFSAdapter.getFileLocation = (config, filename) => {
+      return config.mount + '/files/' + config.applicationId + '/' + encodeURIComponent(filename);
+    }
     const filesController = new FilesController(gridFSAdapter);
-    const result = filesController.expandFilesInObject(config, function () {});
+    const result = await filesController.expandFilesInObject(config, function () { });
```
Comment (mtrezza) [inline spec/FilesController.spec.js:31]: To make sure it all still works, we shouldn't touch any existing tests but only add new ones. After all, this is a feature addition, so any new tests for the new feature (asynchronous invocation) should be needed. The exception being tests for internal methods. Could you revert changes to existing tests?

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: spec/FilesController.spec.js (hunk 2 above)
Comment (vahidalizad) [reply spec/FilesController.spec.js:31]: The expandFilesInObject function depends on getFileLocation, which now supports asynchronous operations. Given this change, I considered two approaches:

1. Check if getFileLocation returns a promise: This would involve adding conditional logic to handle both synchronous and asynchronous scenarios. However, to maintain consistency, this conditional logic would need to be propagated throughout the entire codebase wherever expandFilesInObject is used. This approach seems overly complex and error-prone.

2. Make expandFilesInObject fully async: By making expandFilesInObject an async function and using await for getFileLocation, we simplify the code and ensure it is consistently handling asynchronous operations. This also means updating any functions that rely on expandFilesInObject to be async as well, which I have done, along with adjusting the relevant tests.

I chose the second approach because it results in cleaner, more maintainable code without the need for extensive conditional logic across the codebase. I understand the concern about modifying existing tests, but given the changes, these updates are necessary to ensure that the tests accurately reflect the new asynchronous behavior.

I'm open to further discussion if you have concerns or alternative suggestions.

Comment (mtrezza) [reply spec/FilesController.spec.js:31]: Yes that makes sense. It's good to refactor and optimize internal methods. Just the existing tests for exposed / public APIs need to stay unchanged.

I briefly checked; looks good. Did you make sure that all methods that you changed to async are properly invoked everywhere in code, like `_UnsafeRestQuery.prototype.runFind `?

File: spec/FilesController.spec.js [hunk 3]
```diff
@@ -11,8 +11,8 @@ const mockAdapter = {
   createFile: () => {
     return Promise.reject(new Error('it failed with xyz'));
   },
-  deleteFile: () => {},
-  getFileData: () => {},
+  deleteFile: () => { },
```
Comment (mtrezza) [inline spec/FilesController.spec.js]: Please revert

File: src/Controllers/FilesController.js [hunk 4]
```diff
@@ -15,7 +15,7 @@ export class FilesController extends AdaptableController {
     return this.adapter.getFileData(filename);
   }
 
-  createFile(config, filename, data, contentType, options) {
+  async createFile(config, filename, data, contentType, options) {
```
Comment (mtrezza) [inline src/Controllers/FilesController.js:18]: Is this an internal or public API?

File: src/Routers/UsersRouter.js [hunk 5]
```diff
@@ -609,7 +609,7 @@ export class UsersRouter extends ClassesRouter {
         const userString = req.auth && req.auth.user ? req.auth.user.id : undefined;
         logger.error(
           `Failed running auth step challenge for ${provider} for user ${userString} with Error: ` +
-            JSON.stringify(e),
```
Comment (mtrezza) [inline src/Routers/UsersRouter.js]: Please revert unrelated change

File: spec/FilesController.spec.js (hunk 3 above)
Comment (vahidalizad) [reply spec/FilesController.spec.js]: Done

File: src/Routers/UsersRouter.js (hunk 5 above)
Comment (vahidalizad) [reply src/Routers/UsersRouter.js]: Done

File: spec/FilesController.spec.js (hunk 2 above)
Comment (vahidalizad) [reply spec/FilesController.spec.js:31]: Yeah, as I searched through the project, I only found four instances in three files: UserRouter.js, RestWrite.js, and RestQuery.js. I ensured that all the lines I changed are already covered by existing tests, as confirmed by Codecov.

There was only one function that wasn't already async: `RestWrite.expandFilesForExistingObjects`. However, that function is only used in the main execute function, which was already async. So, everything should work seamlessly.

**Review by mtrezza — APPROVED** [review]
Looks good, let's wait for CI to pass...

**Comment by vahidalizad** [comment]
> Looks good, let's wait for CI to pass...

Could you run that failed test again? I believe it should pass, as the failure seems unrelated to this PR.

**Comment by vahidalizad** [comment]
Great, now that everything is settled, when do you think we can release this? I'd like to get back to the AWS S3 Adapter PR: [refactor: Migrate S3 Client from AWS SDK v2 to v3](https://github.com/parse-community/parse-server-s3-adapter/pull/220).

**Comment by vahidalizad** [comment]
> Looks good, let's wait for CI to pass...

Great, now that everything is settled, when do you think we can release this? I'd like to get back to the AWS S3 Adapter PR: [refactor: Migrate S3 Client from AWS SDK v2 to v3](https://github.com/parse-community/parse-server-s3-adapter/pull/220).

**Comment by mtrezza** [comment]
Sorry for the delay here; this is good to merge, we just need to wait because the alpha branch is currently locked. I hope we can merge on the weekend.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.3.0-alpha.7](https://github.com/parse-community/parse-server/releases/tag/7.3.0-alpha.7)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.3.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.3.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.3.0](https://github.com/parse-community/parse-server/releases/tag/7.3.0)
